WarehouseManagement.Core/Services/DeliveryService.cs

```diff
@@ -296,4 +297,121 @@ public async Task<ICollection<DeliveryDto>> GetAllDeletedAsync()
 
         return deliveries;
     }
+
+    public async Task<DeliveryHistoryDto> GetHistoryAsync(int id)
+    {
+        var delivery = await repository.GetByIdAsync<Delivery>(id);
+
+        if (delivery == null)
+        {
+            throw new KeyNotFoundException($"{DeliveryWithIdNotFound} {id}");
+        }
+
+        var relatedEntriesIds = repository
+            .AllReadOnly<Entry>()
+            .Where(e => e.DeliveryId == delivery.Id)
+            .Select(e => e.Id);
+
+        var changes = repository
+            .AllReadOnly<EntityChange>()
+            .Where(change => int.Parse(change.EntityId) == delivery.Id || relatedEntriesIds.Any(id => id == int.Parse(change.EntityId)));
+
+        var deliveryHistory = new DeliveryHistoryDto
+        {
+            Id = delivery.Id,
+            Changes = await changes.Select(change => new DeliveryHistoryDto.Change
+            {
+                EntityId = int.Parse(change.EntityId),
+                PropertyName = change.PropertyName,
+                From = change.OldValue!,
+                To = change.NewValue!,
+                Type = change.EntityName == "Delivery" ? DeliveryHistoryChangeType.Delivery : DeliveryHistoryChangeType.Entry,
+                LogType = change.PropertyName == "Status" && change.EntityName == "Entry" ? LogType.EntryStatusChange :
+                    change.PropertyName == "Status" && change.EntityName == "Delivery" ? LogType.DeliveryStatusChange :
+                    change.PropertyName == "ZoneId" ? LogType.ZoneChange : LogType.Split // May be refactored based on the Move functionality
+            }).ToListAsync()
+        };
+
+        return deliveryHistory;
+    }
+
+    public async Task ChangeDeliveryStatusIfNeeded(int id)
+    {
+        var delivery = await repository.GetByIdAsync<Delivery>(id);
+
+        if (delivery == null)
+        {
+            throw new KeyNotFoundException($"{DeliveryWithIdNotFound} {id}");
+        }
+
+        DeliveryStatus expectedStatus;
+
+        if (delivery.Entries.All(e => e.StartedProccessing == null))
+        {
+            expectedStatus = DeliveryStatus.Waiting;
+        }
+        else if (delivery.Entries.Any(e => e.StartedProccessing != null) &&
+            delivery.Entries.Any(e => e.FinishedProccessing == null))
+        {
+            expectedStatus = DeliveryStatus.Processing;
+        }
+        else if (delivery.Entries.All(e => e.FinishedProccessing != null))
+        {
+            expectedStatus = DeliveryStatus.Finished;
+        }
+        else
+        {
+            throw new InvalidOperationException("The delivery entries are in an invalid state.");
+        }
```

Comment:
and you can also put these if checks into a method so now this function reads much easier and it would look cleaner.

```
GetDeliveryById
GetExpectedStatus
if(status != expected)
    update
```

Reply:
Refactoring looks good, are you sure that I should also add `GetDeliveryById()`? If so, it should be also a private method and won't be used outside of the service.

![image](https://github.com/InterCarsBulgaria-TechLeadAcademy/warehouse.management.be/assets/91427186/f979edf8-186d-4649-b878-541619f812ee)

Reply:
it can make sense, especially if we can also reuse it in other places